src/addons/messagelog/src/main/java/ee/ria/xroad/proxy/messagelog/LogRecordManager.java

```diff
@@ -43,212 +48,235 @@
-     * @param session the Hibernate session
-     * @param logRecord the log record to save
-     * @throws Exception if an error occurs while communicating with database
+     * @param session the Hibernate session.
+     * @param logRecord the log record to save.
+     * @throws Exception if an error occurs while communicating with database.
      */
-    protected void save(Session session, LogRecord logRecord) {
+    static void save(Session session, LogRecord logRecord) {
         log.trace("save({})", logRecord.getClass());
 
         session.save(logRecord);
     }
 
     /**
      * Associates each log record with the time-stamp record.
-     * @param session the Hibernate session
-     * @param messageRecords the message records
-     * @param timestampRecord the time-stamp record
-     * @param hashChains the time-stamp hash chains
-     * @throws Exception if an error occurs while communicating with database
+     * @param session the Hibernate session.
+     * @param messageRecords the message records.
+     * @param timestampRecord the time-stamp record.
+     * @param hashChains the time-stamp hash chains.
+     * @throws Exception if an error occurs while communicating with database.
      */
-    protected void setMessageRecordsTimestamped(Session session,
-            Long[] messageRecords, TimestampRecord timestampRecord,
-            String[] hashChains) {
+    private static void setMessageRecordsTimestamped(Session session, Long[] messageRecords,
+            TimestampRecord timestampRecord, String[] hashChains) {
         if (log.isTraceEnabled()) {
-            log.trace("setMessageRecordsTimestamped({}, {})",
-                    Arrays.toString(messageRecords),
-                    timestampRecord.getId());
+            log.trace("setMessageRecordsTimestamped({}, {})", Arrays.toString(messageRecords), timestampRecord.getId());
         }
 
         if (hashChains != null && messageRecords.length != hashChains.length) {
-            throw new RuntimeException(
-                    "Must have hash chain for each log record");
+            throw new RuntimeException("Must have hash chain for each log record");
         }
 
+        // Let's perform directly JDBC related work for bulk update.
+        // Needs to flush the session to get access to previously saved timestamp record.
+        session.flush();
+        session.doWork(connection -> setMessageRecordsTimestamped(messageRecords, timestampRecord, hashChains,
+                connection, getConfiguredBatchSize(session)));
+    }
+
+    private static void setMessageRecordsTimestamped(Long[] messageRecords, TimestampRecord timestampRecord,
+            String[] hashChains, Connection connection, int batchSize) throws SQLException {
+        log.trace("setMessageRecordsTimestamped({})", messageRecords.length);
+
+        int storedCount = 0;
+        PreparedStatement stmt = connection.prepareStatement("UPDATE logrecord SET timestamprecord = ?, "
```

Comment:
This line is raising a blocker level bug in Sonarqube: `Use try-with-resources or close this "PreparedStatement" in a "finally" clause.`

Reply:
True. I commit this fix into the branch estonian-6.17.0-1 and made new pull request Estonian-6.17.0-3.